File: backend/src/main/java/com/example/contact/application/service/DeleteContactService.java
package com.example.contact.application.service;

import com.example.contact.application.port.in.DeleteContactUseCase;
import com.example.contact.domain.exception.ContactNotFoundException;
import com.example.contact.domain.port.ContactRepository;
import com.example.contact.domain.valueobject.ContactId;

/**
 * Deletes a contact by id when it exists.
 */
public final class DeleteContactService implements DeleteContactUseCase {

    private final ContactRepository contactRepository;

    public DeleteContactService(ContactRepository contactRepository) {
        this.contactRepository = contactRepository;
    }

    @Override
    public void execute(ContactId id) {
        if (contactRepository.findById(id).isEmpty()) {
            throw new ContactNotFoundException(id);
        }
        contactRepository.delete(id);
    }
}

File: frontend/src/contact/application/service/DeleteContactService.ts
import { BaseMutationService } from '@/shared/application/service/BaseMutationService';
import { ContactId } from '../../domain/valueobject/ContactId';
import { ContactGateway } from '../../domain/port/ContactGateway';
import { DeleteContactUseCase } from '../port/in/DeleteContactUseCase';

export class DeleteContactService extends BaseMutationService<ContactId, void> implements DeleteContactUseCase {
  constructor(private readonly gateway: ContactGateway) {
    super();
  }

  protected async handleMutation(id: ContactId): Promise<void> {
    await this.gateway.delete(id);
  }
}
